core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1351,23 +1352,22 @@ private Scope createScopeForCommonTableExpression(Table table, Optional<Scope> s
             Optional<List<Identifier>> columnNames = withQuery.getColumnNames();
             if (columnNames.isPresent()) {
                 // if columns are explicitly aliased -> WITH cte(alias1, alias2 ...)
+                checkState(columnNames.get().size() == queryDescriptor.getVisibleFieldCount(), "mismatched aliases");
                 ImmutableList.Builder<Field> fieldBuilder = ImmutableList.builder();
-
-                int field = 0;
-                for (Identifier columnName : columnNames.get()) {
-                    Field inputField = queryDescriptor.getFieldByIndex(field);
-                    fieldBuilder.add(Field.newQualified(
-                            QualifiedName.of(table.getName().getSuffix()),
-                            Optional.of(columnName.getValue()),
-                            inputField.getType(),
-                            false,
-                            inputField.getOriginTable(),
-                            inputField.getOriginColumnName(),
-                            inputField.isAliased()));
-
-                    field++;
+                Iterator<Identifier> aliases = columnNames.get().iterator();
```

Comment:
Could you add test?

Reply:
Sure, however I don't know how to do it. Is there any known table having a hidden column somewhere "in the middle"?

Reply:
Could you have a synthetic table like that? You could maybe use dummy connector as: `io.trino.execution.TestCoordinatorDynamicFiltering.TestConnector`

Reply:
When the hidden column is moved to the first position in the table, this fix is covered by a preexisting test in `TestWith` class.

Reply:
Why the test did not fail then?